shell/platform/tizen/channels/app_control_channel.cc

```diff
@@ -0,0 +1,303 @@
+        RegisterReplyHandler(std::move(events));
+        return nullptr;
+      },
+      [this](const EncodableValue* arguments)
+          -> std::unique_ptr<StreamHandlerError<>> {
+        UnregisterReplyHandler();
+        return nullptr;
+      });
+
+  reply_channel_->SetStreamHandler(std::move(reply_channel_handler));
+}
+
+AppControlChannel::~AppControlChannel() {}
+
+void AppControlChannel::NotifyAppControl(void* app_control) {
+  app_control_h clone = nullptr;
+  app_control_h handle = static_cast<app_control_h>(app_control);
+  AppControlResult ret = app_control_clone(&clone, handle);
+  if (!ret) {
+    FT_LOG(Error) << "Could not clone app control: " << ret.message();
+    return;
+  }
+  auto app = std::make_shared<AppControl>(clone);
+  if (!event_sink_) {
+    queue_.push(app);
+    FT_LOG(Info) << "EventChannel not set yet.";
+  } else {
+    SendAppControlDataEvent(app);
+  }
+  map_.insert({app->GetId(), app});
+}
+
+void AppControlChannel::HandleMethodCall(
+    const MethodCall<EncodableValue>& method_call,
+    std::unique_ptr<MethodResult<EncodableValue>> result) {
+  const auto arguments = method_call.arguments();
+  const auto& method_name = method_call.method_name();
+
+  // AppControl is not needed.
+  if (method_name.compare("create") == 0) {
+    CreateAppControl(arguments, std::move(result));
+    return;
+  }
+
+  // AppControl is needed.
+  auto app_control = GetAppControl(arguments);
+  if (app_control == nullptr) {
+    result->Error("Could not find app_control", "Invalid id provided");
+    return;
+  }
+
+  if (method_name.compare("dispose") == 0) {
+    Dispose(app_control, std::move(result));
+  } else if (method_name.compare("reply") == 0) {
+    Reply(app_control, arguments, std::move(result));
+  } else if (method_name.compare("sendLaunchRequest") == 0) {
+    SendLaunchRequest(app_control, arguments, std::move(result));
+  } else if (method_name.compare("setAppControlData") == 0) {
+    SetAppControlData(app_control, arguments, std::move(result));
+  } else if (method_name.compare("sendTerminateRequest") == 0) {
+    SendTerminateRequest(app_control, arguments, std::move(result));
+  } else {
+    result->NotImplemented();
+  }
+}
+
+void AppControlChannel::RegisterEventHandler(
+    std::unique_ptr<EventSink<EncodableValue>> events) {
+  event_sink_ = std::move(events);
+  SendAlreadyQueuedEvents();
+}
+
+void AppControlChannel::UnregisterEventHandler() {
+  event_sink_.reset();
+}
+
+void AppControlChannel::SendAlreadyQueuedEvents() {
+  while (!queue_.empty()) {
+    SendAppControlDataEvent(queue_.front());
+    queue_.pop();
+  }
+}
+
+void AppControlChannel::RegisterReplyHandler(
+    std::unique_ptr<EventSink<EncodableValue>> events) {
+  reply_sink_ = std::move(events);
+}
+
+void AppControlChannel::UnregisterReplyHandler() {
+  reply_sink_.reset();
+}
+
+template <typename T>
+bool AppControlChannel::GetValueFromArgs(const EncodableValue* args,
```

Comment:
It has the same purpose as GetValueFromEncodeableMap in platform_view_channel.cc 
I'd like to ask for a little refactoring so that all tizen channels can use this template in common.

https://github.com/flutter-tizen/engine/blob/681e6d0a92425ee1c5bd643f9de1fdb4a1545de0/shell/platform/tizen/channels/platform_view_channel.cc#L22

Reply:
I've created an issue to handle it separately - https://github.com/flutter-tizen/engine/issues/165 as it will 'touch' other channels' code.
Is it acceptable to you?

Reply:
of course!